src/status_im/ens/core.cljs

```diff
@@ -0,0 +1,163 @@
+
+(defn assoc-state-for [db username state]
+  (assoc-in db [:ens/registration :states username] state))
+
+(defn assoc-details-for [db username k v]
+  (assoc-in db [:ens/registration :details username k] v))
+
+(defn assoc-username-candidate [db username]
+  (assoc-in db [:ens/registration :username-candidate] username))
+
+(defn empty-username-candidate [db] (assoc-username-candidate db ""))
+
+(fx/defn set-state
+  {:events [:ens/set-state]}
+  [{:keys [db]} username state]
+  {:db (assoc-state-for db username state)})
+
+(defn- on-resolve [registry custom-domain? username address public-key s]
+  (cond
+    (= (ethereum/normalized-address address) (ethereum/normalized-address s))
+    (resolver/pubkey registry (name custom-domain? username)
+                     (fn [ss]
+                       (if (= ss public-key)
+                         (re-frame/dispatch [:ens/set-state username :connected])
+                         (re-frame/dispatch [:ens/set-state username :owned]))))
+
+    (and (nil? s) (not custom-domain?)) ;; No address for a stateofus subdomain: it can be registered
+    (re-frame/dispatch [:ens/set-state username :registrable])
+
+    :else
+    (re-frame/dispatch [:ens/set-state username :unregistrable])))
+
+(fx/defn register-name
+  [{:keys [db] :as cofx} contract custom-domain? username address public-key]
+  (let [{:keys [x y]} (ethereum/coordinates public-key)]
+    (wallet/eth-transaction-call
+     cofx
+     {:contract   (contracts/get-address db :status/snt)
+      :method     "approveAndCall(address,uint256,bytes)"
+      :params     [contract
+                   (money/unit->token 10 18)
+                   (abi-spec/encode "register(bytes32,address,bytes32,bytes32)"
+                                    [(ethereum/sha3 username) address x y])]
+      :to-name    "Stateofus registrar"
+      :amount     (money/bignumber 0)
+      :gas        (money/bignumber 200000)
+      :on-result  [:ens/save-username-and-navigate-back custom-domain? username]
+      :on-error   [:ens/on-registration-failure]})))
+
+(defn- valid-custom-domain? [username]
+  (and (ens/is-valid-eth-name? username)
+       (stateofus/lower-case? username)))
+
+(defn- valid-username? [custom-domain? username]
+  (if custom-domain?
+    (valid-custom-domain? username)
+    (stateofus/valid-username? username)))
+
+(defn- state [custom-domain? username]
+  (cond
+    (string/blank? username) :initial
+    (> 4 (count username)) :too-short
+    (valid-username? custom-domain? username) :valid
+    :else :invalid))
+
+(fx/defn set-username-candidate
+  {:events [:ens/set-username-candidate]}
+  [{:keys [db]} custom-domain? username]
+  (let [state  (state custom-domain? username)
+        valid? (valid-username? custom-domain? username)]
+    (merge
+     {:db (-> db
+              (assoc-username-candidate username)
+              (assoc-state-for username state))}
+     (when-let [name (name custom-domain? username)]
```

Comment:
It's a bit confusing that `name` is both a `let` binding and fn name.

Reply:
This `when-let` binding can be moved to `let` immediately above, so that `when` below becomes `(when (and name (=:valid state))`

Reply:
:+1: